inst/include/Rcpp/sugar/functions/strings/trimws.h

```diff
@@ -0,0 +1,184 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// trimws.h: Rcpp R/C++ interface class library -- trimws
+//
+// Copyright (C) 2017 Nathan Russell
+//
+// This file is part of Rcpp.
+//
+// Rcpp is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Rcpp is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Rcpp.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef Rcpp__sugar__trimws_h
+#define Rcpp__sugar__trimws_h
+
+#include <string>
+#include <cstring>
+
+namespace Rcpp {
+namespace sugar {
+namespace detail {
+
+
+/*  NB: std::isspace is not used because it also counts
+    '\f' and '\v' as whitespace, whereas base::trimws only
+    checks for ' ', '\t', '\r', and '\n'  */
+inline bool isws(const char c) {
+    return c == ' ' || c == '\t' || c == '\n' || c == '\r';
+}
+
+inline const char* trim_left(const char* str) {
+    if (!str) {
+        return "";
+    }
+
+    while (isws(*str)) {
+        ++str;
+    }
+
+    return str;
+}
+
+inline const char* trim_right(const char* str) {
+    static std::string buff;
```

Comment:
You make a good point here. I was primarily looking to avoid repeatedly allocating and deallocating memory for a new `std::string` on each function call, which I assumed would happen if had used a buffer with automatic storage duration. However, I think I can just create a (non-`static`) variable in the enclosing function (`trimws`) and pass it by reference to `trim_right` and `trim_both` so that the memory gets released when `trimws` returns.